doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,501 @@
+
+#### Changes to `Key`
+
+For the `Key`, we need to extract everything for the data and name into their own structs.
+This is done for memory-management reasons, as we need to track how many keys point to the same data and/or name.
+
+```c
+struct _KeyData {
+    union {
+        char * c;
+        void * v;
+    } data;
+
+    size_t dataSize;
+
+    uint16_t refs;
+};
+
+struct _KeyName {
+    char * key;
+    size_t keySize;
+
+    char * ukey;
+    size_t keyUSize;
+
+    uint16_t refs;
+};
+
+struct _Key {
+    struct _KeyData * keyData;
+    struct _KeyName * keyName;
+    KeySet * meta;
+    keyflag_t flags;
+
+    uint16_t refs;
+};
+```
+
+The following calculations are based on the AMD64 platform:
+
+In the data structures above, an empty key does have
+
+- with `_keyData` and `_keyName` allocated: 96 bytes.
+- without `_keyData` but with `_keyName`: 72 bytes.
+- with `_keyData` and `_keyName` `NULL`: 32 bytes.
+
+An empty key of the current implementation has 64 bytes.
+
+A copied, non-modified key with the data structure above does always have 32 bytes.
+A copied, non-modified key of the current implementation has at least 64 bytes (for an empty key), but in reality much more as the name and the data are also copied.
+
+As an example, let's take `user:/hosts/ipv6/example.com = 2606:2800:220:1:248:1893:25c8:1946`.
+
+For a single key, it would take for the COW implementation:
+
+`sizeof(_Key) + sizeof(_KeyName) + sizeof(_KeyData) + name + 1 /* \0 */ + ukey + value + 1 /* \0 */` =
+32 + 40 + 24 + 28 + 1 + 25 + 34 + 1 = 185 bytes.
+
+For two keys (one original and one copy), it would take for the COW implementation:
+
+`sizeof(_Key)*2 + sizeof(_KeyName) + sizeof(_KeyData) + name + 1 /* \0 */ + ukey + value + 1 /* \0 */` =
+32\*2 + 40 + 24 + 28 + 1 + 25 + 34 + 1 = 217 bytes.
+
+For a single key, it would take the current implementation:
+
+`sizeof(_Key) + name + 1 + ukey + value + 1` =
+64 + 28 + 1 + 25 + 34 + 1 = 153 bytes
+
+For two keys, the current implementation would take:
+
+`(sizeof(_Key) + name + 1 + ukey + value + 1) * 2` =
+(64 + 28 + 1 + 25 + 34 + 1) \* 2 = 306 bytes.
+
+#### Changes to `KeySet`
+
+For `KeySet`, we need to split out everything to do with the stored keys into a separate datastructure.
+This includes the array itself, the sizes and the hashmap.
+
+Why don't we just add the number of references to the original `KeySet`?
+
+- If we delete a copied KeySet, we don't know which KeySet is the original, so we couldn't decrement the counter.
+  This could be dealt with storing a pointer to the original KeySet.
+- If the original KeySet is deleted, we don't know which other KeySets point at the data, so updating their count would not work
+- In similar fashion, if you update the original KeySet, the copied KeySets will also contain the new data (if the memory address does not change).
+  This is unexpected behaviour.
+
+```c
+struct _KeySetData {
+    struct _Key ** array;
```

Comment:
While we're making these fundamental changes...

One thing I was curious about, but I thought these kind of changes were off the table:

```c
struct _KeySetData {
    struct _KeyName ** array;
    struct _KeyData ** array;

    size_t keySize;
    size_t keyDataSize;

    [...]
}
```

In that case we could do the `mmap` pointer update using vector extensions and probably speed it up significantly.

Reply:
Yeah, I've also though about similarly radical changes in the past, but I've never thought we could actually agree to do something as big as that.

Reply:
Big changes that'll be done in the near future by you before 1.0 are totally possible.

As performance is the lowest of our 4 goals, and memory usage was in the last benchmark a **much** bigger problem than CPU, I don't think a decision that has a focus to speed without a clear usability benefit, has any chance.

If the change is an improvement for e.g. the API user and the speed-up is only a side effect, I cannot wait for your proposal in form of a decision PR.

Reply:
> Big changes that'll be done in the near future by you before 1.0 are totally possible.

Not the ones I've thought of... Even if I had the time to implement them. Basically, I've again and again come to the conclusion that our biggest limitation is the separate `Key` structure (or at least the fact that it is public API). More specifically, that `Key` is actually a whole (key,value) pair despite its name.

For example a [Radix tree](https://en.wikipedia.org/wiki/Radix_tree) (or [HAT-trie](https://en.wikipedia.org/wiki/HAT-trie)) would be prefect for `KeySet`, but it just doesn't work. You'd effectively store the name twice, because you still need the full name in `Key`.

> I think this proposal might break COW for keys.

The much bigger problem here (as alluded to above) is that there is no `Key *` in the sturcture. This immediately makes this snippet impossible

```c
Key * k = keyNew ("user:/foo", KEY_END);
ksAppendKey (ks, k);
assert (k == ksLookupByName (ks, "/foo", 0));
```

More generally: How would all the `ks*` functions that return a `Key *` work at all?
You'd have to allocate a new `Key *` or take a `Key * dest` as argument or use some fixed `Key *` that is only valid until the next call. All of which would be API changes so big and complex that you'd almost have to rewrite the whole codebase and definitely not something that could be automated.